mne/time_frequency/tfr.py

```diff
@@ -197,55 +200,46 @@ def _cwt_fft(X, Ws, mode="same", decim=1):
     fsize = 2 ** int(np.ceil(np.log2(size)))
 
     # precompute FFTs of Ws
-    fft_Ws = np.empty((n_freqs, fsize), dtype=np.complex128)
+    if use_fft:
+        fft_Ws = np.empty((n_freqs, fsize), dtype=np.complex128)
     for i, W in enumerate(Ws):
         if len(W) > n_times:
             raise ValueError('Wavelet is too long for such a short signal. '
                              'Reduce the number of cycles.')
-        fft_Ws[i] = fftn(W, [fsize])
-
-    for k, x in enumerate(X):
-        if mode == "full":
-            tfr = np.zeros((n_freqs, fsize // decim), dtype=np.complex128)
-        elif mode == "same" or mode == "valid":
-            tfr = np.zeros((n_freqs, n_times // decim), dtype=np.complex128)
-
-        fft_x = fftn(x, [fsize])
-        for i, W in enumerate(Ws):
-            ret = ifftn(fft_x * fft_Ws[i])[:n_times + W.size - 1]
-            if mode == "valid":
-                sz = abs(W.size - n_times) + 1
-                offset = (n_times - sz) / 2
-                this_slice = slice(offset // decim, (offset + sz) // decim)
-                tfr[i, this_slice] = _centered(ret, sz)[::decim]
-            else:
-                tfr[i, :] = _centered(ret, n_times)[::decim]
-        yield tfr
+        if use_fft:
+            fft_Ws[i] = fftn(W, [fsize])
 
+    # Decimating is performed after centering the convolution, and can
+    # therefore lead to 1 time sample jittering.
+    jitter = (n_times % decim) % 2
+    n_times_out = fsize if mode == "full" else n_times
+    n_times_out = n_times_out // decim + jitter
 
-def _cwt_convolve(X, Ws, mode='same', decim=1):
-    """Compute time freq decomposition with temporal convolutions
-    Return a generator over signals.
-    """
-    X = np.asarray(X)
-
-    n_signals, n_times = X.shape
-    n_freqs = len(Ws)
-
-    # Compute convolutions
+    # Make generator looping across signals
     for x in X:
-        tfr = np.zeros((n_freqs, n_times // decim), dtype=np.complex128)
+        tfr = np.zeros((n_freqs, n_times_out), dtype=np.complex128)
```

Comment:
Are we guaranteed that the same set of samples will always be written to on each loop iteration? If so, this declaration can go outside the loop.